### PR TITLE
fix: crd definition problems

### PR DESCRIPTION
**Pull request by cd1989**
**What this PR does / why we need it**:

Fix crd definition problems.

**Which issue(s) this PR is related to** *(optional, link to 3rd issue(s))*:

Fixes #

Reference to #

**Special notes for your reviewer**:

/cc @zhujian7 

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/apis/cyclone/v1alpha1/common.go [hunk 1]
```diff
@@ -18,3 +18,9 @@ type ParameterItem struct {
 
 // ArgumentValue defines a argument value
 type ArgumentItem ParameterItem
+
+// KeyValue defines a key-value pair
+type KeyValue struct {
+	Key   string `json:"key"`
+	Value string `json:"key"`
```
Comment (zhujian7) [inline pkg/apis/cyclone/v1alpha1/common.go]: value?

**Comment by ddysher** [comment]
/hold
/bark

I'll take a look at API change.

File: pkg/apis/cyclone/v1alpha1/workflow_trigger.go [hunk 2]
```diff
@@ -27,10 +30,12 @@ const (
 
 // WorkflowTriggerSpec defines workflow trigger definition.
 type WorkflowTriggerSpec struct {
-	// Name of the workflow to trigger
-	WorkflowRef string `json:"workflowRef"`
+	// Reference of the Workflow to trigger
+	WorkflowRef corev1.ObjectReference `json:"workflowRef"`
```
Comment (ddysher) [inline pkg/apis/cyclone/v1alpha1/workflow_trigger.go]: This should be a pointer so you can tell nil reference.

File: pkg/apis/cyclone/v1alpha1/workflow.go [hunk 3]
```diff
@@ -28,3 +28,10 @@ type StageItem struct {
 	// Stages that this stage depends on
 	Depends []string `json:"depends"`
 }
+
```
Comment (ddysher) [inline pkg/apis/cyclone/v1alpha1/workflow.go]: :point_up_2: 

```go
// WorkflowSpec defines workflow specification.
type WorkflowSpec struct {
	Stages StageItem `json:"stageItem"`
}
```
should be:

```go
// WorkflowSpec defines workflow specification.
type WorkflowSpec struct {
	Stages []StageItem `json:"stageItem"`
}
```
?

Comment (ddysher) [reply pkg/apis/cyclone/v1alpha1/workflow.go]: Also, can we change json tag `stageItem` to just `stages`?

File: pkg/apis/cyclone/v1alpha1/stage.go [hunk 4]
```diff
@@ -14,17 +14,17 @@ type Stage struct {
 	metav1.ObjectMeta `json:"metadata,omitempty"`
 	// Stage specification without using template
```
Comment (ddysher) [inline pkg/apis/cyclone/v1alpha1/stage.go]: This comment is incorrect.

File: pkg/apis/cyclone/v1alpha1/stage_template.go [hunk 5]
```diff
@@ -79,5 +79,12 @@ type StageTemplateSpec struct {
 	// Pod kind workload
 	Pod PodWorkload `json:"pod,omitempty"`
 	// CRD kind workload
-	CRD CRDWorkload `json:"crd:omitempty"`
+	CRD CRDWorkload `json:"crd,omitempty"`
```
Comment (ddysher) [inline pkg/apis/cyclone/v1alpha1/stage_template.go]:
```go
Arguments []Argument `json:"arguments,omitempty"`
```

Argument & ArgumentItem is a bit duplicated, can we merge them?  This makes the overall structures easier to reason about.

File: pkg/apis/cyclone/v1alpha1/resource.go [hunk 6]
```diff
@@ -7,7 +7,7 @@ import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
 // Resource represents a resource used in workflow
-type Resource struct {
+type CycloneResource struct {
```
[the remark this answers is not in the document]
Comment (zhujian7) [reply pkg/apis/cyclone/v1alpha1/resource.go]: I had thought that The `Resource` conflict with the *method* `Resource` in `registry.go`.
But maybe the method `Resource` can be removed, I will try again.

**Comment by zhujian7** [comment]
Does all CRDList(WorkflowList, StageList, etc) definition need an annotation:
```
// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
```
to automaticly generate `DeepCopyObject` func.

File: pkg/apis/cyclone/v1alpha1/workflow.go (hunk 3 above)
Comment (cd1989) [reply pkg/apis/cyclone/v1alpha1/workflow.go]:
```
type WorkflowSpec struct {
	Stages []StageItem `json:"stages"`
}
```

You are right.

File: pkg/apis/cyclone/v1alpha1/stage_template.go (hunk 5 above)
Comment (cd1989) [reply pkg/apis/cyclone/v1alpha1/stage_template.go]: `Argument ` defines argument, with `default` value, `description`. And `ArgumentItem` specify value for a argument, `ArgumentValue` may be a better choice.

**Comment by zhujian7** [comment]
I will merge this first since some work depends on it. And we will submit another PR if there is something need to be fixed.

**Pull request merged**